Sources/AsyncAlgorithms/Deadline.swift

```diff
@@ -0,0 +1,60 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct TimeoutError<C: Clock>: Error {
+  public let deadline: C.Instant
+  public let clock: C
+  
+  public init(_ deadline: C.Instant, _ clock: C) {
+    self.deadline = deadline
+    self.clock = clock
+  }
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public func withDeadline<C: Clock, T: Sendable>(
+  _ deadline: C.Instant,
+  clock: C,
+  _ operation: @Sendable () async throws -> T
+) async throws -> T {
+  return try await withoutActuallyEscaping(operation) { operation in
+    try await withThrowingTaskGroup(of: T.self) { group in
+      group.addTask(operation: operation)
+      group.addTask {
+        try await Task.sleep(until: deadline, clock: clock)
+        throw TimeoutError(deadline, clock)
+      }
+      defer { group.cancelAll() }
+      return try await group.next()!
+    }
+  }
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public func withTimeout<C: Clock, T: Sendable>(
+  in duration: C.Duration,
```

Comment:
I don't think such should be called "timeout", rather this still forms _a deadline_ and should be called `with...Deadline(in: .seconds...)` because it is defining a _deadline_ _in_ some seconds, it is not a timeout "in".

Reply:
isn't a deadline in 5 seconds just a timeout starting at now in 5 seconds?

Reply:
I agree here with @ktoso that we should define this with the term `deadline`. The difference between a deadline and a timeout is quite important. A deadline can be reused and it will always be the same point in time w.r.t. the clock used. A timeout would need to recalculated against your current time. Furthermore, a deadline is less error prone since suspended tasks/threads won't influence the actual point in time calculation.